x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -657,6 +660,95 @@ public void getUserPrivileges(AuthorizationInfo authorizationInfo, ActionListene
+                indicesPrivileges.addAll(toIndicesPrivileges(indicesGroup));
+            }
+        }
+
+        listener.onResponse(
+            new RoleDescriptorsIntersection(
+                List.of(
+                    Set.of(
+                        new RoleDescriptor(
+                            UUIDs.randomBase64UUID(),
+                            null,
+                            indicesPrivileges.toArray(new RoleDescriptor.IndicesPrivileges[0]),
+                            null,
+                            null,
+                            null,
+                            null,
+                            null
+                        )
+                    )
+                )
+            )
+        );
+    }
+
+    /**
+     * This may return more than a single index privilege for the indices group because an index privilege can contain at most one query
+     * and FLS group definition, whereas an indices group can contain multiple. We therefore "unfold" the group into multiple privileges,
+     * if necessary.
+     */
+    private static List<RoleDescriptor.IndicesPrivileges> toIndicesPrivileges(final IndicesPermission.Group indicesGroup) {
+        final Set<BytesReference> queries = indicesGroup.getQuery() == null ? Collections.emptySet() : indicesGroup.getQuery();
+        final Set<FieldPermissionsDefinition.FieldGrantExcludeGroup> fieldGrantExcludeGroups = getFieldGrantExcludeGroups(indicesGroup);
+
+        // We need to force at least one iteration even if there is no FLS/DLS, so we use Collections.singleton(null) in the loops below
+        // if either FLS or DLS is not specified
+        final Set<BytesReference> queriesOrNullSingleton = queries.isEmpty() ? Collections.singleton(null) : queries;
+        final Set<FieldPermissionsDefinition.FieldGrantExcludeGroup> fieldGrantExcludeGroupsOrNullSingleton = fieldGrantExcludeGroups
+            .isEmpty() ? Collections.singleton(null) : fieldGrantExcludeGroups;
+        final List<RoleDescriptor.IndicesPrivileges> indicesPrivileges = new ArrayList<>();
+        for (BytesReference query : queriesOrNullSingleton) {
+            for (FieldPermissionsDefinition.FieldGrantExcludeGroup fieldGrantExcludeGroup : fieldGrantExcludeGroupsOrNullSingleton) {
+                final RoleDescriptor.IndicesPrivileges.Builder builder = RoleDescriptor.IndicesPrivileges.builder()
+                    .indices(indicesGroup.indices())
+                    .allowRestrictedIndices(indicesGroup.allowRestrictedIndices())
+                    .privileges(indicesGroup.privilege().name())
+                    .query(query);
+                if (fieldGrantExcludeGroup != null) {
+                    builder.grantedFields(fieldGrantExcludeGroup.getGrantedFields())
+                        .deniedFields(fieldGrantExcludeGroup.getExcludedFields());
+                }
+                indicesPrivileges.add(builder.build());
+            }
+        }
```

Comment:
I am not sure about this block of code. It has O(n^2) complexity, e.g. if there are 10 indices groups in the original role descriptor with DLS and FLS, it will be rewritten to be 100 indices groups. Maybe we should not collate remote indices privileges at the index level as you suggested earlier. What do you think?

Reply:
That's true, but on the other hand this means that there must have been on the order of 100 index privileges when the role was defined to begin with (since that's the only way we would end up with an index group with many queries/FLS definitions that require unfolding), so we would end up with roughly as many operations here, even if we didn't collate (correct me if I'm wrong!).

Collating leaves us with a more compact representation, a more sane output for `GetUserPrivileges`, and combines definitions with the same indices but different privileges into one so it still optimizes slightly. It's also nice that the behavior around remote indices is consistent with regular indices (i.e., we collate for both).

My intuition is that a role with this many queries and FLS definitions is likely a corner case and as such I wonder if it makes sense to go with the current implementation, and then revisit how we handle collation as a future optimization. LMKWYT!

Reply:
We discussed this and @ywangd is right, we end up sending (quadratically) more privileges than necessary in the case when we have a role that has both FLS and DLS. In particular, given a definition like:

```
indices: [
 {"names": ["a"], "privileges": ["read"], "query": "q1", "field_security": {f1} },
 {"names": ["a"], "privileges": ["read"], "query": "q2", "field_security": {f2} },
]
```

we end up sending:

```
{"names": [a], privileges: [read], query: q1, field_security: f1},
{"names": [a], privileges: [read], query: q1, field_security: f2},
{"names": [a], privileges: [read], query: q2, field_security: f1},
{"names": [a], privileges: [read], query: q2, field_security: f2}
```

Two of the above index privileges are redundant. We only need to send privileges such that each query and group is present at least once, i.e., we can send:

```
{"names": [a], privileges: [read], query: q1, field_security: f1},
{"names": [a], privileges: [read], query: q2, field_security: f2}
```

or

```
{"names": [a], privileges: [read], query: q1, field_security: f2},
{"names": [a], privileges: [read], query: q2, field_security: f1}
```

I found a way to optimize the algorithm to only send a number of privileges that is `O(max(n, m))` where `n` is number of queries and `m` is the number of FLS groups, so linear instead of quadratic. I've pushed the changes.

However, this improved algorithm has its own downsides. It's not straightforward to grok, and non-deterministic (because it iterates over unordered sets under the hood), i.e., it may combine DLS and FLS privileges differently on different runs. This is both a problem for debugging and possibly for caching. Making it deterministic is possible but makes things more complicated still.

I think the pragmatic solution here is to _not_ collate remote index privileges; in this case, we will be guaranteed that each indices permission group will only have a single DLS/FLS definition and we do not need to unfold anymore. The benefits of collating don't seem to justify the overall complexity here.